tsfc/kernel_interface/common.py

```diff
@@ -213,6 +213,7 @@ def compile_gem(self, ctx):
         else:
             return_variables = []
             expressions = []
+        expressions = constant_fold_zero(expressions)
```

Comment:
I'm not sure if this is the right place to put this, but `mode.flatten` (more specifically `remove_componenttensor()` somehow generates `Literal([0.])` in complex mode, which loopy eventually complains about (https://github.com/firedrakeproject/firedrake/runs/7659503437?check_suite_focus=true).

Reply:
> more specifically remove_componenttensor() somehow generates Literal([0.]) in complex mode

Is that something we want to happen ? If not, should we not nip the problem in the bud or am I missing something here ?

Reply:
I suspect something was previously relying on the eager conversion. I think it is fine to do this here before we hand off to loop.